MatrixSDK/Data/RoomList/Common/MXRecentsRoomListDataFetcher.swift

```diff
@@ -0,0 +1,158 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+@objcMembers
+internal class MXRecentsRoomListDataFetcher: NSObject, MXRoomListDataFetcher {
+    internal var fetchOptions: MXRoomListDataFetchOptions
+    private weak var session: MXSession?
+    private var recentsRooms: [String] = []
+    
+    private var allRoomSummaries: [MXRoomSummaryProtocol] = []
+    private var sessionSyncObserver: Any?
+    
+    private var recentRoomIds: [String] = []
+
+    private let multicastDelegate: MXMulticastDelegate<MXRoomListDataFetcherDelegate> = MXMulticastDelegate()
+
+    var data: MXRoomListData? {
+        didSet {
+            guard data != nil else {
+                //  do not notify when stopped
+                return
+            }
+            
+            notifyDataChange()
+        }
+    }
+    
+    internal init(fetchOptions: MXRoomListDataFetchOptions,
+                  session: MXSession?) {
+        self.fetchOptions = fetchOptions
+        self.session = session
+        super.init()
+        self.fetchOptions.fetcher = self
+        self.refresh()
+        self.addDataObservers()
+    }
+
+    //  MARK: - Delegate
+    
+    func addDelegate(_ delegate: MXRoomListDataFetcherDelegate) {
+        multicastDelegate.addDelegate(delegate)
+    }
+    
+    func removeDelegate(_ delegate: MXRoomListDataFetcherDelegate) {
+        multicastDelegate.removeDelegate(delegate)
+    }
+    
+    func removeAllDelegates() {
+        multicastDelegate.removeAllDelegates()
+    }
+    
+    internal func notifyDataChange() {
+        multicastDelegate.invoke({ $0.fetcherDidChangeData(self, totalCountsChanged: true) })
+    }
+
+    func paginate() {
+        // Do nothing. We don't paginate breadcrumbs
+    }
+    
+    func resetPagination() {
+        // Do nothing. We don't paginate breadcrumbs
+    }
+    
+    // MARK: - Public
+    
+    func refresh() {
+        guard let events = session?.accountData.accountData["events"] as? [[AnyHashable: Any]] else {
+            MXLog.error("[MXRecentsRoomListDataFetcher] cannot retrieve events")
+            return
+        }
+        
+        var breadcrumbs: [AnyHashable: [String]]?
+        for event in events {
+            if let type = event["type"] as? String,
+                type == kMXAccountDataTypeBreadcrumbs,
+                let content = event["content"] as? [AnyHashable: [String]] {
+                breadcrumbs = content
+            }
+        }
+        
+        guard var recentRoomIds = breadcrumbs?[kMXAccountDataTypeRecentRoomsKey] else {
+            MXLog.error("[MXRecentsRoomListDataFetcher] cannot retrieve recent rooms")
```

Comment:
I wonder if this (and the one above should be `warn` instead of `error` from the perspective of whether its the kind of thing we want to report to Sentry? Presumably every new account would do this.